testsuite/tests/typing-local/local.ml

```diff
@@ -0,0 +1,325 @@
+[%%expect{|
+val local_closure : unit -> unit = <fun>
+|}]
+
+(*
+ * Higher order functions, with arguments that promise not to leak
+ *)
+
+let use_locally (f : ('a[@stack]) -> 'a) (x : 'a) = f x
+(* This version also promises not to leak the closure *)
+let use_locally' : (('a [@stack]) -> 'a [@stack]) -> 'a -> 'a = fun f x -> f x
+[%%expect{|
+val use_locally : (('a [@stack]) -> 'a) -> 'a -> 'a = <fun>
+val use_locally' : (('a [@stack]) -> 'a [@stack]) -> ('a -> 'a [@stackret]) =
+  <fun>
+|}]
+
+let no_leak = use_locally (fun x -> 1) 42
+let no_leak' = use_locally' (fun x -> 1) 42
+[%%expect{|
+val no_leak : int = 1
+val no_leak' : int = 1
+|}]
+
+let leak_id =
+  use_locally (fun x -> x) 42
+[%%expect{|
+Line 2, characters 24-25:
+2 |   use_locally (fun x -> x) 42
+                            ^
+Error: The value x is local, so cannot be used here as it might escape
+|}]
+
+let leak_ref =
+  let r = ref None in
+  use_locally (fun x -> r.contents <- Some x; x) 42
+
+[%%expect{|
+Line 3, characters 43-44:
+3 |   use_locally (fun x -> r.contents <- Some x; x) 42
+                                               ^
+Error: The value x is local, so cannot be used here as it might escape
+|}]
+
+let leak_ref_2 =
+  let[@stack] r = ref None in
+  use_locally (fun x -> let[@stack] _ = r in r.contents <- Some x; x) 42
+[%%expect{|
+Line 3, characters 40-41:
+3 |   use_locally (fun x -> let[@stack] _ = r in r.contents <- Some x; x) 42
+                                            ^
+Error: The value r is local, so cannot be used inside a closure that might escape
+|}]
+
+let leak_ref_3 =
+  let[@stack] r = ref None in
+  use_locally' (fun x -> let[@stack] _ = r in r.contents <- Some x; x) 42
+[%%expect{|
+Line 3, characters 65-66:
+3 |   use_locally' (fun x -> let[@stack] _ = r in r.contents <- Some x; x) 42
+                                                                     ^
+Error: The value x is local, so cannot be used here as it might escape
+|}]
+
+
+
+let no_leak_exn =
+  use_locally (fun x -> let[@stack] _exn = Invalid_argument x in "bluh") "blah"
+[%%expect{|
+val no_leak_exn : string = "bluh"
+|}]
+let do_leak_exn =
+  use_locally (fun x -> let[@stack] _exn = raise (Invalid_argument x) in "bluh") "blah"
+
+[%%expect{|
+Line 2, characters 67-68:
+2 |   use_locally (fun x -> let[@stack] _exn = raise (Invalid_argument x) in "bluh") "blah"
+                                                                       ^
+Error: The value x is local, so cannot be used here as it might escape
+|}]
+
+(* same, but this time the function is allowed to return its argument *)
+let use_locally (f : ('a[@stack]) -> ('a[@stackret])) : ('a[@stack]) -> ('a[@stackret]) = f
+[%%expect{|
+val use_locally :
+  (('a [@stack]) -> ('a [@stackret])) -> ('a [@stack]) -> ('a [@stackret]) =
+  <fun>
+|}]
+
+let loc = ((fun x -> x) : (int[@stack]) -> (int[@stackret]))
+
+let no_leak_id = let[@stack] _ = use_locally ((fun x -> x) : (int[@stack]) -> (int[@stackret])) 42 in ()
+
```

Comment:
Presumably, this one would work if the return value was itself annotated with `[@stack]`. It would be good to include that example too.

Reply:
It will be the fun rather than the return value getting annotated, but yes